### PR TITLE
www-servers/lighttpd: lighttpd-9999.ebuild

### PR DESCRIPTION
**Pull request by gstrauss**
lighttpd-9999.ebuild

Follow-up to #35790

* create lighttpd-9999.ebuild
* USE_LIBDEFLATE flag
   (TODO: drop ~s390 from ebuild KEYWORDS; file rekeywording bug)
* USE_CRYPTO_* flags to select crypto library used by lighttpd
   (What should happen to USE_NETTLE flag?)
* use upstream mime.conf (large list of mime-types)

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [ ] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

### CODE REVIEW

**Comment by gstrauss** [comment]
This is a first attempt to address the questions raised in #35790.  Please advise on the gentoo/ebuild way to do things if there are better idioms.  Thanks!

**Comment by gstrauss** [comment]
The mime.conf file distributed with upstream is generated from `/etc/mime.types` on Fedora.  I think this is fine for a fairly large list of mime types.  If you prefer to generate from Gentoo `/etc/mime.types`, then we should modify the ebuild to install `doc/scripts/create-mime.conf.pl` and run that in the `lighttpd.conf-r3` with `include_shell /etc/lighttpd/create-mime.conf.pl`.  If we do that, I would prefer `/etc/lighttpd/scripts/` as the script installation location, as `doc/scripts/cert-staple.sh` might also be installed there.

**Comment by gstrauss** [comment]
@thesamesam: not urgent -- this is a follow-up to #35790  Would you subscribe to or assign this PR?  Thanks.

**Comment by thesamesam** [comment]
> The mime.conf file distributed with upstream is generated from /etc/mime.types on Fedora. I think this is fine for a fairly large list of mime types. If you prefer to generate from Gentoo /etc/mime.types, then we should modify the ebuild to install doc/scripts/create-mime.conf.pl and run that in the lighttpd.conf-r3 with include_shell /etc/lighttpd/create-mime.conf.pl. If we do that, I would prefer /etc/lighttpd/scripts/ as the script installation location, as doc/scripts/cert-staple.sh might also be installed there.

I agree that this sounds fine to me -- but maybe add a comment as to this effect within the ebuild, given it's such a complicated topic and to note we have options in the event of user complaints?

> USE_CRYPTO_* flags to select crypto library used by lighttpd (What should happen to USE_NETTLE flag?)

This is always tricky -- we debated it a lot over the years for `net-misc/curl` which has a similar kind of concept with the backends it supports vs. the defaults.

I'm not against the state of it as you've proposed, but I wonder if we need it. Is there a compelling reason for someone to not just configure `lighttpd.conf` appropriately to select their chosen backend?

**Comment by gstrauss** [comment]
> > The mime.conf file distributed with upstream is generated from /etc/mime.types on Fedora. I think this is fine for a fairly large list of mime types. If you prefer to generate from Gentoo /etc/mime.types, then we should modify the ebuild to install doc/scripts/create-mime.conf.pl and run that in the lighttpd.conf-r3 with include_shell /etc/lighttpd/create-mime.conf.pl. If we do that, I would prefer /etc/lighttpd/scripts/ as the script installation location, as doc/scripts/cert-staple.sh might also be installed there.
> 
> I agree that this sounds fine to me -- but maybe add a comment as to this effect within the ebuild, given it's such a complicated topic and to note we have options in the event of user complaints?

I updated `www-servers/lighttpd/files/conf/lighttpd.conf-r3` but fear it is too wordy
```
# {{{ includes
# Map file extensions to mime types for Content-Type using mime.conf
# shipped with lighttpd code, based on Fedora /etc/mime.types mapping.
# Option: comment out on embedded systems to save some memory and use
# lighttpd built-in defaults, which include a subset of common web mime types.
# Option: use lighttpd doc/scripts/create-mime.conf.pl to generate mapping
# from Gentoo /etc/mime.types at lighttpd startup.
include "/etc/lighttpd/mime.conf"
# }}}
```

> > USE_CRYPTO_* flags to select crypto library used by lighttpd (What should happen to USE_NETTLE flag?)
> 
> This is always tricky -- we debated it a lot over the years for `net-misc/curl` which has a similar kind of concept with the backends it supports vs. the defaults.
> 
> I'm not against the state of it as you've proposed, but I wonder if we need it. Is there a compelling reason for someone to not just configure `lighttpd.conf` appropriately to select their chosen backend?

These are build options which affect linking, not something which can be a choice in lighttpd.conf.  These options might be used in embedded systems where both on-disk size of binaries and memory use of the libraries matter.  (For most people, this is a minor, unnecessary optimization since lighttpd and supporting libraries are already smaller than many other binaries on their system.)

Is this something desirable in gentoo?  If so, what is the best way to do it?

For reference, openwrt provides these options in openwrt `make menuconfig`
```
  LIGHTTPD_CRYPTOLIB_NETTLE:nettle
  LIGHTTPD_CRYPTOLIB_GNUTLS:gnutls
  LIGHTTPD_CRYPTOLIB_MBEDTLS:mbedtls
  LIGHTTPD_CRYPTOLIB_OPENSSL:openssl
  LIGHTTPD_CRYPTOLIB_WOLFSSL:wolfssl
```
https://github.com/openwrt/packages/blob/master/net/lighttpd/Makefile

**Comment by thesamesam** [comment]
> These are build options which affect linking, not something which can be a choice in lighttpd.conf. These options might be used in embedded systems where both on-disk size of binaries and memory use of the libraries matter. (For most people, this is a minor, unnecessary optimization since lighttpd and supporting libraries are already smaller than many other binaries on their system.)

I'd put it as a comment in the ebuild in `src_configure`.

The rationale being: we _do_ have people who use Gentoo to build embedded systems, but they're used to checking out the relevant packages, and we have ways for users to inject custom build options with e.g. `MYMESONARGS` as an environment variable.

**Comment by gstrauss** [comment]
> I'd put it as a comment in the ebuild in src_configure.

```
+       # choose crypto library
+       # (one specific library might be preferred on embedded systems)
```

Other suggestions welcome.

**Comment by thesamesam** [comment]
How about:
```
+       # choose crypto library
+       # (one specific library might be preferred on embedded systems via MYMESONARGS with e.g. -DBLAH_FORCE)
```

**Comment by gstrauss** [comment]
Modified comment (sorry for delay; needed to sleep):
```
        # choose crypto library
        # (one specific library might be preferred on embedded systems via
        #  MYMESONARGS with e.g. -DFORCE_blah_CRYPTO)
```

---

Is there anything I need to do for the automated build error for this PR?
`9999	SourcingError	failed sourcing ebuild: error sourcing ebuild`
https://qa-reports.gentoo.org/output/gentoo-ci/06d2575ec2/output.html#www-servers/lighttpd

The ebuild is not going to find version 9999 (lighttpd-9999.tar.xz)

File: www-servers/lighttpd/lighttpd-9999.ebuild [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+	local c_defs
+	local with_nettle
+
+	# choose crypto library
+	# (one specific library might be preferred on embedded systems via
+	#  MYMESONARGS with e.g. -DFORCE_blah_CRYPTO)
+	if use crypto-nettle ; then
+		with_nettle=$(meson_use crypto-nettle with_nettle)
+	else if use nettle ; then
```
Comment (thesamesam) [inline www-servers/lighttpd/lighttpd-9999.ebuild:123]: I think this will fix the sourcing error:
```suggestion
	elif nettle ; then
```

Comment (gstrauss) [reply www-servers/lighttpd/lighttpd-9999.ebuild]: doh!  thank you!  fixed.

Comment (gstrauss) [reply www-servers/lighttpd/lighttpd-9999.ebuild]: That was it.  Incorrect shell syntax.  Sorry.  Thanks for diagnosing that.

The CI report is now clean for this PR.

**Comment by thesamesam** [comment]
Thanks! This looks good to me now. I think I'll drop the crypto-* flags now we have the comment, and see if anyone requests them.
